src/ucp/core/ucp_listener.c

```diff
@@ -121,44 +109,77 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
     return (elem->cb == ucp_listener_conn_request_progress) && (listener == arg);
 }
 
-static ucs_status_t ucp_listener_conn_request_callback(void *arg,
-                                                       const void *conn_priv_data,
-                                                       size_t length)
+static ucs_status_t
+ucp_listener_ep_conn_create(ucp_listener_h listener, void *id,
+                            const ucp_wireup_sockaddr_priv_t *client_data,
+                            size_t length, ucp_listener_accept_t *accept)
+{
+    accept->listener     = listener;
+    accept->is_ep        = 0;
+    accept->conn_request = ucs_malloc(ucs_offsetof(ucp_conn_request_t,
+                                                   priv_addr) +
+                                      length, "accept connection request");
+    if (accept->conn_request == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    accept->conn_request->listener = listener;
+    accept->conn_request->id       = id;
+    memcpy(&accept->conn_request->priv_addr, client_data, length);
+
+    return UCS_OK;
+}
+
+static void
+ucp_listener_conn_request_callback(uct_iface_h tl_iface, void *arg,
+                                   uct_conn_request_h conn_request,
+                                   const void *conn_priv_data, size_t length)
 {
-    ucp_listener_h        listener = arg;
+    ucp_listener_h listener                       = arg;
+    const ucp_wireup_sockaddr_priv_t *client_data = conn_priv_data;
+    uct_worker_cb_id_t prog_id                    = UCS_CALLBACKQ_ID_NULL;
     ucp_listener_accept_t *accept;
-    uct_worker_cb_id_t    prog_id;
+    ucp_worker_h worker;
+    ucs_status_t status;
 
     ucs_trace("listener %p: got connection request", listener);
 
     /* Defer wireup init and user's callback to be invoked from the main thread */
-    accept = ucs_malloc(sizeof(*accept), "ucp_listener accept");
-    if (accept == NULL) {
-        ucs_error("failed to allocate listener accept context");
-        return UCS_ERR_NO_MEMORY;
-    }
+    accept = ucs_malloc(sizeof(*accept), "ucp_listener_accept");
```

Comment:
looks like we allocate both 'accept' and 'conn_request', can 'conn_request' be inside 'accept' and the use container_of, or some other way to allocate only once and not twice?

Reply:
No, because `ucp_conn_request_t` is not completed type. See
https://github.com/evgeny-leksikov/ucx/pull/3/files#diff-328d160be0fceb7e1c3a7d1f022c2ef3R321